src/compiler/checker.ts

```diff
@@ -1329,11 +1329,14 @@ export const enum CheckMode {
     Inferential = 1 << 1,                           // Inferential typing
     SkipContextSensitive = 1 << 2,                  // Skip context sensitive function expressions
     SkipGenericFunctions = 1 << 3,                  // Skip single signature generic functions
-    IsForSignatureHelp = 1 << 4,                    // Call resolution for purposes of signature help
-    RestBindingElement = 1 << 5,                    // Checking a type that is going to be used to determine the type of a rest binding element
+    SkipReturnTypeFromBodyInference = 1 << 4,       // Skip inferring from return types of context sensitive functions

```

Comment:
It's separate from `SkipGenericFunctions` because `getReturnTypeFromBody` *removes* `SkipGenericFunctions` bit - so when the compiler gets to a function node with type parameters within a context-sensitive function's return it can't just use `SkipGenericFunctions` to skip over inferring its own return type from body. 

The goal of `SkipReturnTypeFromBodyInference` is to persist through `getReturnTypeFromBody`.